### PR TITLE
MES-7804 Implement unsuccessful tests with 16 driving faults

### PR DESCRIPTION
**Pull request by gabrielngdvsa**
## Description

* PLEASE INSERT DESCRIPTION

## Checklist

- [x] PR title includes the JIRA ticket number and an understandable description
- [x] Branch is rebased against the latest develop
- [x] Code has been tested manually
- [x] PR link added to JIRA ticket

### CODE REVIEW

**Review by solomonsimukoko — APPROVED** [review]

File: src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature [hunk 1]
```diff
@@ -0,0 +1,62 @@
+Feature: Cat-B Unsuccessful feature
+
+  Scenario:   Successful test with 16 faults
```
Comment (matthew2564) [inline src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature:3]: Should this be ```Unsuccessful```?

File: src/pageobjects/des/finalOutcomeScreen.pageobject.ts [hunk 2]
```diff
@@ -128,6 +133,63 @@ class FinalOutcomePageObject extends Page {
...
+              case 'yes':
+                await clickElement('click', 'selector', this.d255YesInput);
+                break;
+              case 'no':
+                await clickElement('click', 'selector', this.d255NoInput);
+                break;
+              default:
+                console.info(`Could not find ${fieldInput}`);
+            }
+            break;
```
Comment (matthew2564) [inline src/pageobjects/des/finalOutcomeScreen.pageobject.ts:159]: Not sure if this is general practice in this repo, although I can't imagine a switch inside a switch is a recommended way of doing things!

Comment (solomonsimukoko) [reply src/pageobjects/des/finalOutcomeScreen.pageobject.ts:159]: @matthew2564  the completePassedFinalOutcomePage has a very similar structure and has a switch as well

Comment (matthew2564) [reply src/pageobjects/des/finalOutcomeScreen.pageobject.ts:159]: Fair enough, although I think this pattern/implementation needs to be reviewed

Comment (gabrielngdvsa) [reply src/pageobjects/des/finalOutcomeScreen.pageobject.ts:159]: thanks matt, just been following whatever convention is, happy to have chat further around this?

File: src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      | faultSelector                                                 | faultExpectedText                | faultCountSelector                                         | numberOfFaults      |
+      | des-debrief-screen::control-accelerator-driver-fault          | Control - Accelerator            | des-debrief-screen::control-accelerator-driver-fault-count | 16                  |
+    And I click on the button "des-debrief-screen::end-debrief-btn"
+    And I click on the element "#continue-to-non-pass-finalisation"
+    And I wait on element "des-final-outcome-screen::page-title-non-pass" to be displayed
+    And I complete the unsuccessful final outcome screen from a database
+      | d255                          | na              |
+      | testLanguage                  | na              |
+      | debriefWitnessed              | yes             |
+      And I pause for 2000ms
```
[the remark this answers is not in the document]
Comment (gabrielngdvsa) [reply src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature:36]: Runs okay without it, will remove thanks

File: src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    And I wait on element "des-final-confirmation-screen::page-title" to be displayed
```
[the remark this answers is not in the document]
Comment (gabrielngdvsa) [reply src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature:37]: Added thanks

File: src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      | faultSelector                                                 | faultExpectedText                | faultCountSelector                                         | numberOfFaults      |
+      | des-debrief-screen::control-accelerator-driver-fault          | Control - Accelerator            | des-debrief-screen::control-accelerator-driver-fault-count | 16                  |
+    And I click on the button "des-debrief-screen::end-debrief-btn"
+    And I click on the element "#continue-to-non-pass-finalisation"
+    And I wait on element "des-final-outcome-screen::page-title-non-pass" to be displayed
+    And I complete the unsuccessful final outcome screen from a database
+      | d255                          | na              |
+      | testLanguage                  | na              |
+      | debriefWitnessed              | yes             |
+    # And I pause for 2000ms
```
Comment (turnerc1) [inline src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature:36]: Remove commented code

**Review by turnerc1 — CHANGES_REQUESTED** [review]

File: src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    And I wait on element "des-waiting-room-to-car::page-title" to be displayed
+    And I complete the waiting room to car page
+    And I wait on element "des-test-report-screen::page-title" to be displayed
+    And I select all legal requirements
+    And I add "16" driving faults to the "des-test-report-screen::competency-btn-accelarator" field
+    And I expect that container "des-test-report-screen::driving-fault-counter" contains the text "16"
+    And I click on the button "des-test-report-screen::end-test-btn-xpath"
+    And I click on the button "des-test-report-screen::end-test-continue-to-debrief-btn"
+    And I wait on element "des-debrief-screen::test-outcome-failed" to be displayed
+    And I check from a database the number of faults are correct
```
Comment (turnerc1) [inline src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature:26]: it still says database

File: src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      | faultSelector                                                 | faultExpectedText                | faultCountSelector                                         | numberOfFaults      |
+      | des-debrief-screen::control-accelerator-driver-fault          | Control - Accelerator            | des-debrief-screen::control-accelerator-driver-fault-count | 16                  |
+    And I click on the button "des-debrief-screen::end-debrief-btn"
+    And I click on the element "#continue-to-non-pass-finalisation"
+    And I wait on element "des-final-outcome-screen::page-title-non-pass" to be displayed
+    And I complete the unsuccessful final outcome screen from a database
```
Comment (turnerc1) [inline src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature:32]: it still says database

File: src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      | d255                          | na              |
+      | testLanguage                  | na              |
+      | debriefWitnessed              | yes             |
+    And I wait on element "des-final-confirmation-screen::page-title" to be displayed
+    And I check the confirmation page details from a database
```
Comment (turnerc1) [inline src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature:37]: it still says database

File: src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      | provLicenceRecieved   | na                                          |
+      | transmission          | na                                          |
+      | d255                  | No                                          |
+    And I click on the button "des-final-confirmation-screen::submit-test-results-btn"
+    And I wait on element "des-final-confirmation-screen::submit-test-modal-title" to be displayed
+    And I click on the button "des-final-confirmation-screen::submit-modal-submit-btn"
+    And I wait on element "des-back-to-office-holding-page::continue-to-write-up-btn" to be displayed
+    And I click on the button "des-back-to-office-holding-page::continue-to-write-up-btn"
+    And I wait on element "des-office-page::page-title" to be displayed
+    And I complete the Office unsuccessful page from a database
```
Comment (turnerc1) [inline src/features/des/cat-b/UnsuccessfulTests/UnsuccessfulTestWith16Faults.feature:50]: it still says database

File: src/steps/des/when.ts [hunk 10]
```diff
@@ -48,6 +50,11 @@ When(
   completePassedFinalOutcomePage,
 );
 
+When(
+  /^I complete the unsuccessful final outcome screen from a database$/,
```
Comment (turnerc1) [inline src/steps/des/when.ts:54]: it still says database

File: src/steps/des/when.ts [hunk 11]
```diff
@@ -63,6 +70,11 @@ When(
   completeOfficePage,
 );
 
+When(
+  /^I complete the Office unsuccessful page from a database$/,
```
Comment (turnerc1) [inline src/steps/des/when.ts:74]: it still says database

File: src/steps/des/when.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (gabrielngdvsa) [reply src/steps/des/when.ts:54]: Following the wording convention , it just means theres a data table in the steps 

 And I complete the Office page from a database
      | activityCode            | NA                                          |
      | routeNumber             | 2                                           |
      | independentDriving      | sat nav                                     |
      | trueLikeness            | yes                                         |

[the remark this answers is not in the document]
Comment (gabrielngdvsa) [reply src/steps/des/when.ts:54]: refactoring per teams chat

**Review by turnerc1 — APPROVED** [review]

**Pull request merged**